### DISCUSSION TITLE
My Wishlist (And probably of some others). I will also try to contribute with the code soon.

### DISCUSSION THREAD

**Discussion by mukundshah** [valinet/ExplorerPatcher#972] [Ideas]
### Taskbar
- Windows 11's default taskbar but with windows icon on the left corner, and icons in the center. However the start menu can be independent of the windows icon position. Simply, I would like to have windows icon on the left corner but the start menu to open in the center.
- When taskbar size is forced to be small for Windows 11 default, the clock and other icons get trimmed down. I think, removing the date from the clock part will fix this. Or some how decreasing the padding of that section and reducing font size of clock and date should also do the job.

### Window switcher
- option to only show app icon and title below it when hovered/highlighted (like macOS app switcher) if `Display only the most recent window (per application switcher)` is set true. And consecutively pressing the `Alt` + `` ` `` will cycle through window of the highlighted/selected app.
- highlighted item in the switcher uses an outline box. I would be better to have something more distinguishable. Like, dimming or darkening the area around it. (Something like this: https://d33wubrfki0l68.cloudfront.net/fda400671d6ef965ab100d691901a97cdbf91f38/83532/public/demo/frontpage.jpg )
- option to show app/window from all desktop or all monitor or all desktop and monitor (related to first one in the list)
- make switcher drag-aware i.e. item from one app can be dragged to another through the switcher, like we did/do with taskbar.
- support for gestures

**Comment by valinet** [comment]
### Taskbar
> - Windows 11's default taskbar but with windows icon on the left corner, and icons in the center. However the start menu can be independent of the windows icon position. Simply, I would like to have windows icon on the left corner but the start menu to open in the center.

How do you plan on implementing that?

> - When taskbar size is forced to be small for Windows 11 default, the clock and other icons get trimmed down. I think, removing the date from the clock part will fix this. Or some how decreasing the padding of that section and reducing font size of clock and date should also do the job.

Again, how do you plan to implement that?

### Window switcher
> - option to only show app icon and title below it when hovered/highlighted (like macOS app switcher) if `Display only the most recent window (per application switcher)` is set true. And consecutively pressing the `Alt` + `` ` `` will cycle through window of the highlighted/selected app.

The cleanest way to handle this is to virtualize the DWM Thumbnails API calls. When in this mode, it will ignore actually displaying the thumbnail (`DwmUpdateThumbnailProperties` becomes an empty stub). When querying the source info size (`DwmQueryThumbnailSourceSize`) in this mode, it should return the dimensions desired for the area where the icon will be drawn. That takes care of the nasty stuff without altering the code too much. Then, it's a matter of having it draw the app icon where the preview would have gone, and the title at the bottom or wherever...

> - highlighted item in the switcher uses an outline box. I would be better to have something more distinguishable. Like, dimming or darkening the area around it. (Something like this: https://d33wubrfki0l68.cloudfront.net/fda400671d6ef965ab100d691901a97cdbf91f38/83532/public/demo/frontpage.jpg )

Current highlighter is borrowed from Windows 10. Yeah, a background could be added, although, meh, it's cosmetics again, I prefer it the way it is. Supporting less options is also easier, but yeah, an option can always be introduced.

> - option to show app/window from all desktop or all monitor or all desktop and monitor (related to first one in the list)

I don't understand what you mean by this. There is already a `Show windows only from current monitor`.

> - make switcher drag-aware i.e. item from one app can be dragged to another through the switcher, like we did/do with taskbar.

And this would help with...? Not to mention it takes a lot of work to implement, a ton of time to test, polish and make it work right, but what would it solve? Use this to arrange windows on the desktop, change their Z-order, which roughly equates to the order in which windows were last used? When would this be useful and for what?

> - support for gestures

Gestures like...? Related to #516?

**Reply by mukundshah** [reply]
> Gestures like...? Related to #516?

Yes!

> I don't understand what you mean by this. There is already a `Show windows only from current monitor`.

Currently, the switcher doesn't show apps from all virtual desktops. Is that not implemented or am I missing something?

**Reply by valinet** [reply]
> Currently, the switcher doesn't show apps from all virtual desktops. Is that not implemented or am I missing something?

Indeed, the switcher displays only windows from the current desktop. Good luck identifying the others. The thing is, it's not that easy to do. Unless there is some API for working with virtual desktops that offers info about this, or again some undocumented API is employed, this is next to impossible to do reliably.

Windows on other virtual desktops are cloaked. That's how you identify them, you use `DwmGetWindowAttribute` for that. Cloaking means the windows are rendered but the compositor does not display them in the final image. The problem is, not only windows on other virtual desktops are cloaked. Any application can cloak its windows at free will and there are many cloaked windows on the current virtual desktop just fine. So yeah, with that, not really any chance. So some other API should be identified and used. I don't use virtual desktops, so I never bothered doing anything regarding them. Do you have any plan here?

> Yes!

Again, good luck with that as well, again a tough one. Afaik, there is no public API for that, the implementation is internal, to Microsoft's discretion. I detailed in that post what the approaches may be. Once you get notifications for fingers up/down/left/right etc, connecting them to work with the switcher would indeed be a breeze. And connecting the switcher to show when the gestures are made should be trivial as well, so long as you get to that point.

**Reply by mukundshah** [reply]
> Once you get notifications for fingers up/down/left/right etc, connecting them to work with the switcher would indeed be a breeze.

[github.com/ModernFlyouts-Community/ModernFlyouts](https://github.com/ModernFlyouts-Community/ModernFlyouts)
They were able to perform the gestures. Maybe you want to look in this., or have you already? Let me know. 
I was eventually planning to use their logic (if possible), before I wrote this discussion.

**Reply by valinet** [reply]
They were able to do what? Do they handle 3/4 finger swipes on the touchpad? Or what gestures did they use? It wouldn't hurt to be a but more verbose. The project's huge, if you know something, you can say rather than speak generally.

**Reply by mukundshah** [reply]
Sorry about not being precise 😅.
They have built custom flyouts for volume, brightness, etc. (which you have became aware of)
At beginning, these custom controls only appeared if interacted with keyboard. If the user has set three finger or four finger up/down gesture to change volume,  then the native Volume HUD (the default black rectangle HUD) appeared. More on this: https://github.com/ModernFlyouts-Community/ModernFlyouts/issues/4

Later they managed to get it work with precision gesture as well. In newer version, the swipe gesture will show the custom HUD.
https://github.com/ModernFlyouts-Community/ModernFlyouts/releases/tag/0.9.3

**Reply by valinet** [reply]
Yeah, well, it's a bit more complicated than that. From what I see, they just watch when the built-in flyout comes up (which it does by the means of gestures), and then hide that and show theirs. Fine. Let's say I watch for the built-in Alt-Tab and then hide it and show mine, hypothetically (it wouldn't work since I need to register the `Alt`+`Tab` hotkey for myself as well and that disables the built-in flyout). But suppose I do that. Then the next problem is I don't know where the swipe was: left/right/up/down. They tell easily because they also get volume up/down notifications (as if keys were pressed), so indirectly they determine the user scrolled up/down. In my case, I still don't have a way to identify where the swipe went.

They also said the things lives in `twinui`/`twinui.pcshell`, which coroborates with what I know. Indeed, it's some interface there, although yeah, it has to be studied and hooked, probably.

So yeah, it's fucked up. I described some avenues of exploitation in the issue posted earlier.

**Comment by Shomnipotence** [comment]
Now we have a Start Button on the corner. And open the Start Menu in the left. And at the same time, the App Icons are in the center right?

**Reply by valinet** [reply]
What?

**Reply by valinet** [reply]
If you are asking about center Windows 10 Start menu, honestly, I am starting to question how some of you think about me: I am not retarded, ofc I felt like having it centered makes sense - I just haven’t had the time to pull it off. When it will be ready, it will be posted…

**Reply by valinet** [reply]
And the reason why existing threads are reused to post totally unrelated questions also is beyond my capacity of understanding…

**Reply by Shomnipotence** [reply]
> And the reason why existing threads are reused to post totally unrelated questions also is beyond my capacity of understanding…

I found it has available now, I should delete this post
